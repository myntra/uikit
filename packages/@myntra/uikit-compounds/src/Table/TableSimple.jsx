import React from 'react'
import PropTypes from 'prop-types'
import { classnames } from '@myntra/uikit-utils'

import Row from './TableRow'
import Cell from './TableCell'

import styles from './Table.module.css'
import Sentinel from './Sentinel'

/* eslint-disable react/prop-types */
const WithSentinel = /* istanbul ignore next: trivial style generation code (difficult to test). */ ({ children }) => (
  <Sentinel
    className={classnames('sentinel').use(styles)}
    transform={({ height }) => ({ marginTop: '-' + (parseFloat(height) - 1) + 'px' })}
    transformSentinel={({ height }) => ({ height: parseFloat(height) - 2 + 'px' })}
  >
    {children}
  </Sentinel>
)
/* eslint-enable react/prop-types */

/**
 A simple table.

 @since 0.3.0
 @status EXPERIMENTAL
 */
function TableSimple({ head, body, layout, className, renderRow: RowWrapper, useDiv }) {
  const Table = useDiv ? 'div' : 'table'
  const Thead = useDiv ? 'div' : 'thead'
  const Tbody = useDiv ? 'div' : 'tbody'
  return (
    <div className={classnames('wrapper').use(styles)}>
      <div className={classnames('scroller').use(styles)}>
        <Table className={classnames(className, 'table', { fixed: layout === 'fixed' }).use(styles)}>
          <Thead className={classnames('thead').use(styles)}>
            {head.rows.map(row => (
              <RowWrapper key={row.id} head>
                <Row useDiv={useDiv}>
                  {row.columns.map(({ children, key, ...props }) => (
                    <Cell {...props} head key={key} useDiv={useDiv}>
                      {children}
                    </Cell>
                  ))}
                </Row>
              </RowWrapper>
            ))}
          </Thead>
          <Tbody className={classnames('tbody').use(styles)}>
            {body.rows.map(row => (
              <RowWrapper key={row.id} data={row.data} index={row.index} WithSentinel={WithSentinel}>
                <Row useDiv={useDiv}>
                  {row.columns.map(({ children, key, ...props }) => (
                    <Cell key={key} {...props} useDiv={useDiv}>
                      {children}
                    </Cell>
                  ))}
                </Row>
              </RowWrapper>
            ))}
          </Tbody>
        </Table>
      </div>
    </div>
  )
}

TableSimple.propTypes = {
  className: PropTypes.string,
  head: PropTypes.object,
  body: PropTypes.object,
  layout: PropTypes.oneOf(['fixed', 'auto']),
  renderRow: PropTypes.func,
  /** Render div with display table. */
  useDiv: PropTypes.bool
}

TableSimple.defaultProps = {
  renderRow({ children }) {
    return React.Children.only(children)
  }
}

export default TableSimple
