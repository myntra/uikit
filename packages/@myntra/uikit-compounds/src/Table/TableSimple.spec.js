import React from 'react'
import { mount } from 'enzyme'
import TableSimple from './TableSimple'

function prepareForTest(props) {
  return mount(
    <TableSimple
      head={{
        colgroup: [{ colSpan: 2, key: 'name', column: {} }],
        rows: [
          {
            id: 0,
            columns: [{ key: 'firstName', children: 'First Name' }, { key: 'lastName', children: 'Last Name' }]
          }
        ]
      }}
      body={{
        rows: [
          {
            id: 1,
            columns: [{ key: 'firstName', children: 'John', head: true }, { key: 'lastName', children: 'Doe' }]
          }
        ]
      }}
      renderRow={({ children }) => children}
      {...props}
    />
  )
}

it('should render correctly', () => {
  const wrapper = prepareForTest()

  expect(wrapper).toMatchSnapshot()
})
