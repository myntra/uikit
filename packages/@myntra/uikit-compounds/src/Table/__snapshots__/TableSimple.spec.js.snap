// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<TableSimple
  body={
    Object {
      "rows": Array [
        Object {
          "columns": Array [
            Object {
              "children": "John",
              "head": true,
              "key": "firstName",
            },
            Object {
              "children": "Doe",
              "key": "lastName",
            },
          ],
          "id": 1,
        },
      ],
    }
  }
  head={
    Object {
      "colgroup": Array [
        Object {
          "colSpan": 2,
          "column": Object {},
          "key": "name",
        },
      ],
      "rows": Array [
        Object {
          "columns": Array [
            Object {
              "children": "First Name",
              "key": "firstName",
            },
            Object {
              "children": "Last Name",
              "key": "lastName",
            },
          ],
          "id": 0,
        },
      ],
    }
  }
  renderRow={[Function]}
>
  <div
    className="wrapper"
  >
    <div
      className="scroller"
    >
      <table
        className="table"
      >
        <thead
          className="thead"
        >
          <renderRow
            head={true}
            key="0"
          >
            <Row>
              <tr
                className="row"
              >
                <Cell
                  head={true}
                  key="firstName"
                >
                  <th
                    className="col head"
                  >
                    First Name
                  </th>
                </Cell>
                <Cell
                  head={true}
                  key="lastName"
                >
                  <th
                    className="col head"
                  >
                    Last Name
                  </th>
                </Cell>
              </tr>
            </Row>
          </renderRow>
        </thead>
        <tbody
          className="tbody"
        >
          <renderRow
            WithSentinel={[Function]}
            key="1"
          >
            <Row>
              <tr
                className="row"
              >
                <Cell
                  head={true}
                  key="firstName"
                >
                  <th
                    className="col head"
                  >
                    John
                  </th>
                </Cell>
                <Cell
                  key="lastName"
                >
                  <td
                    className="col"
                  >
                    Doe
                  </td>
                </Cell>
              </tr>
            </Row>
          </renderRow>
        </tbody>
      </table>
    </div>
  </div>
</TableSimple>
`;

exports[`should use div 1`] = `
<TableSimple
  body={
    Object {
      "rows": Array [
        Object {
          "columns": Array [
            Object {
              "children": "John",
              "head": true,
              "key": "firstName",
            },
            Object {
              "children": "Doe",
              "key": "lastName",
            },
          ],
          "id": 1,
        },
      ],
    }
  }
  head={
    Object {
      "colgroup": Array [
        Object {
          "colSpan": 2,
          "column": Object {},
          "key": "name",
        },
      ],
      "rows": Array [
        Object {
          "columns": Array [
            Object {
              "children": "First Name",
              "key": "firstName",
            },
            Object {
              "children": "Last Name",
              "key": "lastName",
            },
          ],
          "id": 0,
        },
      ],
    }
  }
  renderRow={[Function]}
  useDiv={true}
>
  <div
    className="wrapper"
  >
    <div
      className="scroller"
    >
      <div
        className="table"
      >
        <div
          className="thead"
        >
          <renderRow
            head={true}
            key="0"
          >
            <Row
              useDiv={true}
            >
              <div
                className="row"
              >
                <Cell
                  head={true}
                  key="firstName"
                  useDiv={true}
                >
                  <div
                    className="col head"
                  >
                    First Name
                  </div>
                </Cell>
                <Cell
                  head={true}
                  key="lastName"
                  useDiv={true}
                >
                  <div
                    className="col head"
                  >
                    Last Name
                  </div>
                </Cell>
              </div>
            </Row>
          </renderRow>
        </div>
        <div
          className="tbody"
        >
          <renderRow
            WithSentinel={[Function]}
            key="1"
          >
            <Row
              useDiv={true}
            >
              <div
                className="row"
              >
                <Cell
                  head={true}
                  key="firstName"
                  useDiv={true}
                >
                  <div
                    className="col head"
                  >
                    John
                  </div>
                </Cell>
                <Cell
                  key="lastName"
                  useDiv={true}
                >
                  <div
                    className="col"
                  >
                    Doe
                  </div>
                </Cell>
              </div>
            </Row>
          </renderRow>
        </div>
      </div>
    </div>
  </div>
</TableSimple>
`;
